python/cog/server/worker.py

```diff
@@ -44,44 +46,164 @@
+    def __init__(
+        self, predictor_ref: str, tee_output: bool = True, concurrent: int = 1
+    ) -> None:
         self._state = WorkerState.NEW
         self._allow_cancel = False
+        self._semaphore = asyncio.Semaphore(concurrent)
+        self._concurrent = concurrent
 
         # A pipe with which to communicate with the child worker.
         events, child_events = _spawn.Pipe()
-        self._events: "AsyncPipe[_PublicEventType]" = AsyncPipe(events)
         self._child = _ChildWorker(predictor_ref, child_events, tee_output)
-        self._terminating = False
+        self._events: "AsyncPipe[tuple[str, _PublicEventType]]" = AsyncPipe(
+            events, self._child.is_alive
+        )
+        # shutdown requested
+        self._shutting_down = False
+        # stop reading events
+        self._terminating = asyncio.Event()
+        self._mux = Mux(self._terminating)
+        self._predictions_in_flight = set()
+
+    # @contextlib.contextmanager
+    # def ctx(self):
+    #     self._events = AsyncPipe(events)
+    #     with self._events.executor:
+    #         self._child = _ChildWorker(predictor_ref, child_events, tee_output)
+    #         yield self
+    #         if self._child.is_alive():
+    #             self._child.terminate()
+    #             self._child.join()
 
     def setup(self) -> AsyncIterator[_PublicEventType]:
         self._assert_state(WorkerState.NEW)
         self._state = WorkerState.STARTING
-        self._child.start()
 
-        return self._wait(raise_on_error="Predictor errored during setup")
+        async def inner() -> AsyncIterator[_PublicEventType]:
+            # in 3.10 Event started doing get_running_loop
+            # previously it stored the loop when created, which causes an error in tests
+            if sys.version_info < (3, 10):
+                self._terminating = self._mux.terminating = asyncio.Event()
+
+            self._child.start()
+            self._ensure_event_reader()
+            async for event in self._mux.read("SETUP", poll=0.1):
+                yield event
+                if isinstance(event, Done):
+                    if event.error:
+                        raise FatalWorkerException(
+                            "Predictor errored during setup: " + event.error_detail
+                        )
+                    self._state = WorkerState.IDLE
+
+        return inner()
+
+    @contextlib.asynccontextmanager
+    async def prediction_ctx(self, input: PredictionInput) -> AsyncIterator[None]:
+        async with self._semaphore:
+            if self._semaphore._value == 0:
+                # maybe this will fix hypothesis
+                self._state = WorkerState.BUSY
+            self._allow_cancel = True
+            self._predictions_in_flight.add(input.id)
+            try:
+                yield
+            finally:
+                self._predictions_in_flight.remove(input.id)
+        if self._semaphore._value == self._concurrent:
```

Comment:
hmm, I'd be fine with that. there's a new approach with moderately less hoop-jumping. this setup seems more legible to hypothesis without being too ugly, though I don't yet know what the final picture will look like for testing (and testing concurrency). the new approach makes predictions_in_flight: set[str] the source of truth and derives state from that. this seems like it could go either way.